### ISSUE TITLE
Add 'Governance' and 'Compliance' object retention modes for aws_s3_bucket

### ISSUE THREAD

**Issue by voroale** [terraform-providers/terraform-provider-aws#6872]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Hi,
As of current moment - [v0.11.11](https://github.com/hashicorp/terraform/releases/tag/v0.11.11) - one cannot simply configure [S3 Object Lock](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock-overview.html) as a part of object `versioning` configuration or otherwise.
Being able to set **Retention Modes** - _Governance_ and _Compliance_ is crucial for some businesses, that are prone to regulatory constraints.

### Potential Terraform Configuration

Something like `retention_mode` in `versioning` section
```hcl
resource "aws_s3_bucket" "versioning_bucket" {
  bucket = "my-versioning-bucket"
  acl    = "private"

  versioning {
    enabled = true
    retention_mode = "governance" // "compliance" , "generic"
  }
}
```

**Comment by ewbankkit** [comment]
Related:
* https://github.com/terraform-providers/terraform-provider-aws/issues/6634

**Comment by ewbankkit** [comment]
It looks like we will need to modify a couple of resources for full support:
* [`aws_s3_bucket`](https://www.terraform.io/docs/providers/aws/r/s3_bucket.html)
* [`aws_s3_bucket_object`](https://www.terraform.io/docs/providers/aws/r/s3_bucket_object.html)

Suggested syntax:

```hcl
resource "aws_s3_bucket" "foo" {
  bucket = "my-versioning-bucket"
  acl    = "private"

  object_lock_configuration {
    object_lock_enabled = true

    rule {
      default_retention {
        mode = "GOVERNANCE"
        days = 30
      }
    }
  }
}
```

```hcl
resource "aws_s3_bucket_object" "foo" {
  bucket = "my-versioning-bucket"
  key    = "important-data.txt"

  retention {
    mode              = "GOVERNANCE"
    retain_until_date = "2020-01-05T00:00:00.000Z"
  }

  legal_hold {
    status = "ON"
  }
}
```

**Comment by bflad** [comment]
Support for S3 object lock configuration in the `aws_s3_bucket` resource has been merged and will release with version 1.56.0 of the Terraform AWS provider, likely in the next few hours.

For tracking `aws_s3_bucket_object` support, its probably best to create a separate feature request issue. 👍

**Comment by bflad** [comment]
This has been released in [version 1.56.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1560-january-16-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
